src/gui/editlocallyhandler.cpp

```diff
@@ -0,0 +1,333 @@
+        qCWarning(lcEditLocallyHandler) << "Could not start token check."
+                                        << "accountState:" << _accountState
+                                        << "relPath:" << _relPath
+                                        << "token:" << _token;
+        return;
+    }
+
+    const auto encodedToken = QString::fromUtf8(QUrl::toPercentEncoding(_token)); // Sanitise the token
+    const auto encodedRelPath = QUrl::toPercentEncoding(_relPath); // Sanitise the relPath
+
+    _checkTokenJob.reset(new SimpleApiJob(_accountState->account(),
+                                          QStringLiteral("/ocs/v2.php/apps/files/api/v1/openlocaleditor/%1").arg(encodedToken)));
+
+    QUrlQuery params;
+    params.addQueryItem(QStringLiteral("path"), prefixSlashToPath(_relPath));
+    _checkTokenJob->addQueryParams(params);
+    _checkTokenJob->setVerb(SimpleApiJob::Verb::Post);
+    connect(_checkTokenJob.get(), &SimpleApiJob::resultReceived, this, &EditLocallyHandler::remoteTokenCheckResultReceived);
+
+    _checkTokenJob->start();
+}
+
+void EditLocallyHandler::remoteTokenCheckResultReceived(const int statusCode)
+{
+    qCInfo(lcEditLocallyHandler) << "token check result" << statusCode;
+
+    constexpr auto HTTP_OK_CODE = 200;
+    _tokenVerified = statusCode == HTTP_OK_CODE;
+
+    if (!_tokenVerified) {
+        showError(tr("Could not validate the request to open a file from server."), tr("Please try again."));
+        return;
+    }
+
+    proceedWithSetup();
+}
+
+void EditLocallyHandler::proceedWithSetup()
+{
+    if (!_tokenVerified) {
+        qCWarning(lcEditLocallyHandler) << "Could not proceed with setup as token is not verified.";
+        return;
+    }
+
+    const auto foundFiles = FolderMan::instance()->findFileInLocalFolders(_relPath, _accountState->account());
+
+    if (foundFiles.isEmpty()) {
+        if (isRelPathExcluded(_relPath)) {
+            showError(tr("Could not find a file for local editing. Make sure it is not excluded via selective sync."), _relPath);
+        } else {
+            showError(tr("Could not find a file for local editing. Make sure its path is valid and it is synced locally."), _relPath);
+        }
+        return;
+    }
+
+    _localFilePath = foundFiles.first();
+    _folderForFile = FolderMan::instance()->folderForPath(_localFilePath);
+
+    if (!_folderForFile) {
+        showError(tr("Could not find a folder to sync."), _relPath);
+        return;
+    }
+
+    const auto relPathSplit = _relPath.split(QLatin1Char('/'));
+    if (relPathSplit.isEmpty()) {
+        showError(tr("Could not find a file for local editing. Make sure its path is valid and it is synced locally."), _relPath);
+        return;
+    }
+
+    _fileName = relPathSplit.last();
+
+    Systray::instance()->destroyEditFileLocallyLoadingDialog();
+    Q_EMIT setupFinished();
+}
+
+QString EditLocallyHandler::prefixSlashToPath(const QString &path)
+{
+    auto slashPrefixedPath = path;
```

Comment:
You don't always have to make a copy of the path. Something like the following should work:
`return path.startsWith('/') ? path : QString('/') + path;`

Reply:
Much neater, thanks